active-rfcs/0000-name-mangling.md

```diff
@@ -0,0 +1,56 @@
+# Name
+
+- Start Date: 2021-12-05
+- RFC PR: [#2](https://github.com/boa-lang/rfcs/pull/2)
+- Implementation PR: (leave this empty)
+<!-- When when linking to the implementation PR, it should be done in the format of boa-lang/<repo>#<PR number> -->
+
+## Summary
+
+Defines how name mangling will work.
+
+## Basic example
+
+A method of the class `foo` with the signature `method(data: int) -> string` in the module `bar.baz` would be mangled to `boad2m3barm3bazc3fooi6methodp1t3i32r5string`
+## Motivation
+
+Most languages "mangle" names of symbols, in order to allow for things like namespacing, overloading, and class/struct methods. This RFC defines how Boa can mangle names
+
+## Detailed design
+
+For purposes of this RFC, "string" refers to a number n followed by n characters.
+Additionally, "constant string" is a literal string of characters.
+
+### Function name mangling
+A function name will be mangled according to the following scheme
+1. Leader: always the constant string `boa`
+2. Module: a mangled module name
+3. Class (optional): the constant string `c` followed by a string representing the class name
+4. Method type (Required if class exists): `i` (for instance methods) or `s` (for static methods)
+5. Function name: a string representing the name of the function in Boa code
+6. Parameters: the constant string `p` followed by a number n, then n Parameter Descriptors
+7. Parameter Descriptors: the constant string `t` followed by a string representing the type name. If the type is a class, the class name is also mangled
+8. Return type: the constant string `r` followed by the type name
+
+### Class name mangling:
+A class name will be mangled according to the following scheme:
+1. Leader: always the constant string `boaclass`
+2. Module: a mangled module name
+3. Class name: the constant string `n` followed by a string representing the class name
+
+### Module name mangling:
+A module name is mangled according to the following scheme
+1. Depth: the contstant string `d` followed by the modules "depth" (henceforth referred to as d) in the module tree. `std` has a depth of `1` and `std.math` has a depth of `2`
+2. Module names: d repetitions of the following: the constant string `m` followed by a string representing the module name
+
+## Drawbacks
+
+- Generating a mangled name can be slightly costly, and can lead to confusing binaries. A possibility to disable mangling for certain symbols should be considered in the future
```

Comment:
that can be fixed by using things like underscores to act as a sort of whitespace

Reply:
Actually, if we choose to go that route, we _cpuld_ use whitespace as whitespace, but it may be a terrible idea

Reply:
I more meant that underscores can be a good replacement for actual spaces where spaces aren't allowed in names.

Reply:
Yeah, I was referring to the fact that symbols in the binary can have any name we want them to

Reply:
underscores def make it nicer, i agree with lyxal here

Reply:
Can you try to find some prior art that uses underscores? All I could find uses RLE.

Reply:
One other thing of note, humans likely wouldn't look at mangled names directly, they'd see it through something like https://github.com/luser/rustfilt